[1단계 - 로또(자동)] 포키(권예진) 미션 제출합니다.

YJGwon opened the pull request:
반갑습니다 제이! `항상 이유를 알고 싶은` 포키입니다.
제가 저번 미션에서 TDD 사이클을 잘 지키지 못해서 이번엔 비즈니스 로직을 구현할 때 TDD를 사이클을 정확하게 따라가 보는 게 가장 큰 목표였어요.
사실 구현을 마감 10분 전에 끝내서, 원시값 포장이나 일급 컬렉션 요구사항을 제대로 지키지 못했어요.
마감이 다가올 수록 작동하는 프로그램을 빠르게 만들어내려고 test code나 요구사항을 제쳐두고 일단! 구현부터.. 하게 되었네요...😥
부족한 코드지만 잘 부탁드립니다!

YJGwon committed:
docs: 기능 목록 최초 작성

야호와 포키가 함께 상의해서 작성

YJGwon committed:
feat: 로또 한 장에 6개의 숫자 포함

임의의 숫자 6개를 저장해두어 추후에 조건에 맞게 리팩터링이 필요

YJGwon committed:
feat: 금액에 맞는 수량만큼 로또 생성

YJGwon committed:
test: Collections.shuffle() 학습용 테스트

YJGwon committed:
feat: 1 ~ 45 사이의 오름차순 랜덤 번호 생성

YJGwon committed:
feat: 투입 금액에 대한 예외 처리 추가

1. 숫자여야 한다
2. 0보다 커야 한다
3. 1000원으로 나누어 떨어져야 한다

YJGwon committed:
feat: 특정 숫자와 당첨 번호의 일치 확인

List.contains() 그대로 사용한 로직이지만, TDD cycle 실행 순서 기록의 의미로 Test code 보존

YJGwon committed:
feat: 당첨 번호와 일치하는 숫자 개수 계산

보너스 번호에 대한 매칭 로직 추가:
   Lotto와 WinningNumbers에 대한 getter 없이 구현

당첨 갯수에 따른 보너스 번호 매칭 호출 구현 필요

YJGwon committed:
feat: 보너스볼 제외한 일치 개수에 따른 등수 반환

YJGwon committed:
feat: 보너스볼 포함한 일치 개수에 따른 등수 반환

YJGwon committed:
feat: 각 등수별 당첨금 계산

Prize.pickAmount(int count): count만큼의 당첨금 반환

YJGwon committed:
feat: 총 당첨금과 수익률 계산

Money.rate(int numerator): 금액 대비 numberator의 비율 계산

YJGwon committed:
refactor: Money.from() try-catch 내에서 바로 return

YJGwon committed:
feat: 하나의 당첨 번호에 대한 예외처리

1. 숫자만 입력
2. 1 이상 45 이하

YJGwon committed:
feat: 당첨 번호에 대한 개수와 중복에 대한 예외처리

YJGwon committed:
feat: 보너스 번호에 대한 예외 처리

처리된 예외
   1. 숫자만 입력
   2. 1 이상 45 이하
interface LottoNumber 추가
   로또 넘버에 공통적으로 적용되는 조건인 1 이상 45 이하에 대한 예외처리

당첨 번호와 보너스 번호의 중복 예외처리 필요

YJGwon committed:
refactor: parameter에 포장된 WinningNumber 받고, 보너스 번호 관리 책임 분리

WinningNumber.equals(), .hashcode() 추가: 포장 후 중복검사 안되는 오류 해결
WinningNumber.match(int number) 추가: 로또에 대한 당첨 개수 검사 로직에 사용
불필요한 BonusNumber 관리 책임을 제거

YJGwon committed:
feat: 보너스 번호와 로또 번호 중복 예외처리

WinningNumbers에서 가지고 있던 보너스 번호 관리 책임 BonusNumber로 이동

YJGwon committed:
feat: 입력 로직 추가

1. 구입 금액
2. 당첨 번호
3. 보너스 번호

YJGwon committed:
refactor: Lottos.purchase()에 사용되는 Money.amount 정보 은닉

Money.purchase(int price): 현재 금액으로 주어진 가격의 물건을 몇 개 살 수 있는지 계산

YJGwon committed:
feat: Application, Controller 추가

Controller에 구입 금액 입력까지 추가

YJGwon committed:
refactor: 구매 수량 계산 로직 Lotto로 이동

Money.purchase() -> Money.countAvailable() 메서드 이름 변경

YJGwon committed:
feat: 구매 수량 출력

YJGwon committed:
fix: Lotto.numbers와 NOMINEE_NUMBERS의 참조 관계 제거

서로 계속 참조하고 있어 만들어진 모든 로또가 같은 번호를 갖고 있던 문제 해결

YJGwon committed:
feat: 생성된 로또들의 번호 출력

YJGwon committed:
fix: 입력값에 포함된 공백 제거

YJGwon committed:
feat: 총 당첨 통계 도출

MatchResult 추가: 맞는 번호 개수와 보너스 당첨 여부를 한 번에 저장
Prize에 NONE 추가: 당첨되지 않을 경우
PrizeInformation에서 해당 Prize의 당첨 개수 도출

PrizeTest 리팩터링 필요(Lotto를 생성한 후, 각 case에 맞게 결과 조작하는 로직 구현 필요)
PrizeInformation의 생성자 private으로 변경 필요

YJGwon committed:
feat: 지난 주 당첨 번호와 보너스 번호 입력

YJGwon committed:
refactor: 패키지 분리

YJGwon committed:
feat: 당첨 통계 출력

PrizeInformationDTO 추가 : prize의 상태와 prize 당첨 개수 view에 전달
Prize.bonus 추가 : 보너스 번호 포함 여부 판별

YJGwon committed:
feat: 수익률 출력, controller 리팩터링

jjj0611 commented on src/main/java/lotto/Controller.java:
```diff
@@ -0,0 +1,76 @@
+
+        WinningNumbers winningNumbers = getWinningNumbers();
+        PrizeInformations prizeInformations =
+                getPrize(lottos, winningNumbers, getBonusNumber(winningNumbers));
```
해당 메서드의 역할이 너무 많은 것 같아요!
프라이즈를 구하는 것까지는 좋은데 프린트하는 로직까지 끼어 있네요!

YJGwon replied:
프린트 하는 로직은 Controller.run()으로 빼주고, Controller 내의 method들 전체적으로 refactoring 해주었습니다!

jjj0611 commented on src/main/java/lotto/Controller.java:
```diff
@@ -0,0 +1,76 @@
+        return purchaseLottos(purchaseCount);
+    }
+
+    private static int getPurchaseCount(Money money) {
```
purchase count를 구하는 듯한 메서드에서 출력을 하고 있고,
lotto에서 1장당 금액을 관리하고 있네요.

메서드와 객체의 역할이 더 잘 나뉘어지면 좋을 것 같아요

YJGwon replied:
Lotto에서는 Money로 몇 장 살수 있는지 계산해주고, Lottos에서는 Money를 받아 알맞은 장 수 만큼 구매하도록 refactoring했습니다.

jjj0611 commented on src/main/java/lotto/model/prize/Prize.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Arrays;
+
+public enum Prize {
+    FIRST(6, 2_000_000_000, false),
+    SECOND(5, 30_000_000, true),
+    THIRD(5, 1_500_000, false),
+    FOURTH(4, 50_000, false),
+    FIFTH(3, 5_000, false),
+    NONE(0, 0, false);
```
_를 표현해서 잘 구현해주셨네요!
enum을 잘 사용해주셨습니다!

jjj0611 reviewed (changes requested):
안녕하세요 포키! 리뷰어 제이입니다.
로또 미션으로 만나서 반가워요! 

tdd를 열심히 하려고 노력했다고 하셨는데, 커밋 로그를 보니까 그게 잘 보이네요 :)
아마 처음에는 많이 힘들고 어려웠을텐데 고생 많으셨어요!

그래서 시간도 많이 부족해서 제출 기한을겨우 지켜 내신 것으로 이해를 했어요!
그래서 일단 리뷰를 많이 남기진 않고 포키가 한 번 리팩토링을 진행해주시면 좋을 것 같아요.

리팩토링의 방향성에 대해서 말씀을 드리면, 지금 객체가 하는 역할이 잘 구분되지 않았다는 생각이 많이 들어요.
1000원에 대한 값이 로또에도 있고, 돈에도 있고, 1과 45라는 값이 로또 넘버에도 있고, 로또라는 것에도 있고, 현재 로또는 로또 넘버를 가지고 쓰는게 아니라 내부적으로 integer를 쓰고 있네요. 이런 부분들이 전체적인 통일성이 조금 아쉽다고 느껴져요! 

리팩토링의 조건으로 
- 1000이라는 숫자는 한곳에 존재한다
- 1과 45라는 숫자는 한 곳에 존재한다
- 6이라는 숫자는 한 곳에 존재한다

적어도 위 3가지 조건을 충족시키면서 리팩토링 해주세요!
혹시 방향성에 대해 고민이 들거나 이해가 가지 않는 부분이 있으면 편하게 DM 주셔도 됩니다!

YJGwon committed:
refactor: 로또에 쓰이는 모든 번호 LottoNumber로 포장

- 로또 번호의 조건을 LottoNumber에서만 검사
- 후보 번호를 세팅하고 랜덤 번호를 추첨하는 로직 LottoNumber로 이동
- Comparable을 상속받아 LottoNumber끼리 값 비교

YJGwon committed:
refactor: 당첨 번호와 보너스 번호를 묶는 WinningLotto 추가

- BonusNumber 삭제: WinningLotto에서 보너스 번호에 대한 검증을 처리할 수 있어 Wrapping 불필요
- MatchResult.of() 삭제: WinningLotto에서 바로 매칭 결과 생성
- 당첨 번호 개수 로직 Lotto -> WinningNumbers로 이동:
   Lotto가 WinningNumbers를 의존하지 않도록

Lotto 생성 시 LottoNumber가 뽑은 번호가 중복되는 문제 해결 필요

YJGwon committed:
refactor: 6개의 LottoNumber를 가지는 일급 컬렉션 LottoNumbers 추가

- WinningNumbers 삭제: 로또 번호 6개 묶음을 하나의 타입으로 통일
- LottoNumber.draw() parameter 추가: 하나씩 뽑지 않고 size 지정 -> 중복  오류 해결

YJGwon committed:
refactor: LottoDTO 1개 생성하는 private 팩토리 메서드 추가

YJGwon committed:
refactor: Lotto 구매에 알맞은 단위 Money는 모르도록 수정

YJGwon committed:
refactor: Lotto와 Lottos가 Money를 바로 받아 구매하도록 수정

YJGwon committed:
refactor: Lotto의 생성자에서 LottoNumber을 받도록 수정

- 로또가 자동일지 수동일지를 Lotto의 생성자가 아닌 Lottos.purchaseAuto()에서 결정
- PrizeTest 사용할 수 있게 수정

YJGwon committed:
refactor: PrizeInformation 생성자 캡슐화

YJGwon committed:
refactor: 모든 Domain Class에 설명 주석 추가

YJGwon commented:
안녕하세요 제이!
먼저 저에게 고민할 시간을 허락해주셔서 너무 감사드립니다.
스스로의 생각을 많이 녹여내지 못한 채 리뷰를 받게 되는게 아쉬웠는데, 덕분에 주말동안 많이 생각해 볼 수 있었어요.

제이가 주신 3가지 힌트를 따라가봤더니, 자연스럽게 같은 속성을 가진 도메인들을 하나의 Type으로 정의하게 되었습니다.
가장 큰 변화는
- 로또에 사용되는 모든 번호를 `LottoNumber`로 포장한 것
- 6개의 `LottoNumber`를 한 단위로 나타낼 수 있는 일급 컬렉션 `LottoNumbers`를 추가한 것
- 당첨 로또 정보를` WinningLotto`로 모은 것
이렇게 세 가지 입니다.

코멘트 남겨주신 부분도 반영하고 해당 코멘트에 댓글 달아두었습니다.
테스트 메서드도 모두 사용 가능하게 수정해두었고, 역할 분담을 명료하게 하는 것을 염두에 두고 리팩터링 해보았어요.

---

그리고 질문이 하나 있습니다!
리팩터링을 하면서 각 도메인의 역할을 스스로 remind 하기 위해 모든 도메인 클래스에 javadoc을 흉내낸 주석을 추가해봤는데요, 

- 실무에서 javadoc을 많이 쓰는지
- 쓴다면 형식을 정확하게 지키는지
- 지금 상태라면 거슬리는 주석일지

궁금해요.
어찌보면 코드와는 큰 관련이 없을 수도 있지만, 이런 애매한게 정말 검색해도 찾기 힘들더라고요.. 그래서 여쭤보고 싶었어요 😓

jjj0611 commented on src/main/java/lotto/model/number/LottoNumber.java:
```diff
@@ -0,0 +1,76 @@
+/*
+ * 로또 번호로 쓸 수 있는 int에 대한 Wrapper Class
+ */
+public class LottoNumber implements Comparable<LottoNumber> {
```
lotto 번호에 대한 역할과 lotto 번호 6개를 생성하는 역할 2가지를 담당하는 것 같아요!
또한, 로또 번호 6개를 만들기 위해 미리 생성해둔 45개의 로또 넘버가 있는데, from이라는 메서드에서는 새로 로또 번호를 생성하네요!
캐싱해둔 친구를 사용하면 어떨까요?

YJGwon replied:
'캐싱'이라는 개념이 아직 제 머릿속에 안착하지 못했나봐요. 전혀 생각하지 못했는데 너무 좋은 방법같아요! lotto 번호 생성의 역할도 좀 더 고민해보겠습니다.

jjj0611 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    public static Money from(String input) {
+        try {
+            return new Money(Integer.parseInt(input));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(ERROR_TYPE);
+        }
+    }
```
정적 팩토리 메서드를 잘 사용해주셨네요!

jjj0611 commented on src/main/java/lotto/model/number/LottoNumbers.java:
```diff
@@ -0,0 +1,81 @@
+                .collect(Collectors.toList());
+    }
+
+    public int getSize() {
+        return this.lottoNumbers.size();
+    }
+
+    public LottoNumber get(int index) {
+        return this.lottoNumbers.get(index);
+    }
```
test에서만 사용하는 메서드인 것 같아요!
왜 필요한 메서드인가요??

YJGwon replied:
test가 아니면 필요하지 않은 메서드가 맞습니다. 심지어 이미 있는 getValues를 사용하면 test에서도 사용하지 않게 되네요! 반영해두겠습니다.

jjj0611 commented on src/main/java/lotto/model/prize/MatchResult.java:
```diff
@@ -0,0 +1,22 @@
+ * 당첨 번호와 구입한 로또의 대조 결과를 담는 Class
+ */
+public class MatchResult {
+    private int count;
+    private boolean bonus;
```
변경되지 않는다면 final을 붙여보는건 어떨까요?

YJGwon replied:
한 번 당첨 번호가 정해지면 변하지 않게 되는 값이니까 final을 붙여야겠네요. 반영하겠습니다.

jjj0611 reviewed (approved):
안녕하세요 포키! 리뷰어 제이입니다!

먼저 요구사항 드린대로 너무 잘 짜와주셨네요!
커밋 라인 좀 따라 갔는데, TDD를 적용하느라 시간이 많이 부족했던 것으로 보이고,
다시 제출 주신 거에서 구조가 눈에 띄게 바뀌었네요 :)
너무 고생하셨습니다!!

간단하게 코멘트 몇개만 드렸고, 2단계 진행하시면서도 구조가 조금 바뀔 수 있을 것 같아서
좀 더 많은 코멘트는 2단계 때 한 번에 드릴게요!

질문 주신 것에 대해서 답변드리면,
제가 속한 팀에서는 굳이 사용하지는 않습니다.
대부분 코드를 보고도 이해할 수 있는 방향으로 짜는 것을 선호합니다.

또한, 로또 같은 작은 애플리케이션에서는 정말 코드만 보고 이해할 수 있는 방향으로 구현하는 연습을 하는게 중요할 것 같아요!
주석을 절대 쓰면 안된다라기보다는 주석을 쓰지 않고도 코드를 보는 사람을 이해시키기 위한 노력을 기울여본다고 생각하시면 좋을 것 같아요 :)

그럼 1단계 너무 고생 많으셨고, 코드는 이만 머지 할게요!
2단계 때 뵈어요!!(당연한 이야기지만 빠르게 요청을 주실수록 더 많은 이야기를 할 수 있습니다 😄)
